src/ft2/_layout.cpp

```diff
@@ -0,0 +1,174 @@
+  FT_Int32 flags,
+  FT_BBox const& bbox) :
+  glyphs{glyphs}, rectangles{rectangles}, flags{flags}, bbox{bbox}
+{}
+
+Layout Layout::simple(
+  std::u32string const& string,
+  FT_Face const& face,
+  FT_Int32 flags)
+{
+  auto has_kerning = FT_HAS_KERNING(face);
+  auto previous = char32_t{};
+  auto pos = FT_Vector{};
+  auto glyphs = std::vector<FT_Glyph>{};
+  glyphs.reserve(string.size());
+  for (auto& codepoint: string) {
+    auto current = FT_Get_Char_Index(face, codepoint);
+    if (has_kerning && previous && current) {
+      auto delta = FT_Vector{};
+      FT_CHECK(
+        FT_Get_Kerning, face, previous, current, FT_KERNING_DEFAULT, &delta);
+      pos.x += delta.x;
+      pos.y += delta.y;
+    }
+    FT_CHECK(FT_Load_Glyph, face, current, flags);
+    auto glyph = FT_Glyph{};
+    FT_CHECK(FT_Get_Glyph, face->glyph, &glyph);
+    FT_CHECK(FT_Glyph_Transform, glyph, nullptr, &pos);
+    glyphs.push_back(glyph);
+    // 16.16 -> 26.6.
+    pos.x += glyph->advance.x >> 10;
+    pos.y += glyph->advance.y >> 10;
+  }
+  return {glyphs, {}, flags, compute_bbox(pos, glyphs, {})};
+}
+
+Layout Layout::manual(
+  std::vector<std::tuple<FT_Glyph, double, double>> const& positioned_glyphs,
+  std::vector<rect_t> const& rectangles,
+  FT_Int32 flags)
+{
+  auto glyphs = std::vector<FT_Glyph>{};
+  glyphs.reserve(positioned_glyphs.size());
+  for (auto& [glyph, x, y]: positioned_glyphs) {
+    auto copy = FT_Glyph{};
+    FT_CHECK(FT_Glyph_Copy, glyph, &copy);
+    auto pos =
+      FT_Vector{FT_Pos(std::round(x * 64)), FT_Pos(std::round(-y * 64))};
+    FT_Glyph_Transform(copy, nullptr, &pos);
+    glyphs.push_back(copy);
+  }
+  return {glyphs, rectangles, flags, compute_bbox({}, glyphs, rectangles)};
+}
+
+Layout::~Layout()
+{
+  if (!moved) {
+    std::for_each(glyphs.begin(), glyphs.end(), FT_Done_Glyph);
+  }
+}
+
+Layout::Layout(Layout&& other) :
+  glyphs{std::move(other.glyphs)},
+  rectangles{std::move(other.rectangles)},
+  flags{std::move(other.flags)},
+  bbox{std::move(other.bbox)},
+  moved{true}
+{}
+
+py::array_t<uint8_t> Layout::render()
+{
+  auto xmin = int(std::floor(bbox.xMin / 64.)),
+       xmax = int(std::ceil(bbox.xMax / 64.)),
+       ymin = int(std::floor(bbox.yMin / 64.)),
+       ymax = int(std::ceil(bbox.yMax / 64.)),
+       width = xmax - xmin,
+       height = ymax - ymin;
+  auto array =
+    py::array_t<uint8_t>{size_t(height * width), nullptr}
+    .attr("reshape")(height, width).cast<py::array_t<uint8_t>>();
+  auto buf = array.mutable_unchecked().mutable_data(0);
+  std::memset(buf, 0, width * height);
+  for (auto glyph: glyphs) {
+    // TODO: Aliased version.
+    if (glyph->format != FT_GLYPH_FORMAT_BITMAP) {
+      // Do *not* destroy the previous glyph: this would invalidate a
+      // Python-level Glyph holding the old value.
+      FT_CHECK(
+        FT_Glyph_To_Bitmap, &glyph, FT_RENDER_MODE_NORMAL, nullptr, false);
+    }
+    auto b_glyph = reinterpret_cast<FT_BitmapGlyph>(glyph);
```

Comment:
strict aliasing

Reply:
This is actually the documented approach: https://www.freetype.org/freetype2/docs/reference/ft2-base_interface.html#FT_Render_Glyph